### PR TITLE
[WIP] Build URLs for Linux Distributions

### PR DESCRIPTION
**Pull request by jlord**
We need to support downloading MongoDB for Linux distributions like RHEL and SuSE.

This is **WIP** cause there's some stuff to verify (below) and I'm testing things in a really slow way, remotely on a RHEL machine. 

- [ ] Use async and remove sync `fs` calls
- [ ] Figure out discrepancies between RHEL 7.0 and `rhel72`
- [ ] Figure out if the Linux urls currently being put together actually exist? 

**NOTE** There are not always existing builds for a version of these distributions, for example, there is no version after `3.6.0-rc0` for `rhel`.

### CODE REVIEW

**Comment by jlord** [comment]
@imlucas 👋 

We're trying to detect Linux distributions and machine architectures for running our tests on Evergreen. Previously here we weren't detecting all the archs and we weren't detecting different Linux distributions. 

I added an option property `linuxDistribution` that auto-detects the distribution (of a few possible ones we know we're testing on) and I updated to check for more kinds of architectures. 

But here in lines the problem. In the tests we're specifying what all the options are and some of them are (a legacy) Linux without a distribution. Previously this was fine because we weren't dealing with distributions, but now (in this PR) we're detecting it so, for instance, when Travis ran the tests it was adding Ubuntu as the distribution to the URLs that the tested wanted to be legacy Linux.

The easiest way to get around that is to specify when a url needs to be created for a legacy Linux and leave off the distribution of the Linux machine that is possibly running this. But this is essentially a breaking change and I wasn't sure if there are places besides this test that are generating legacy Linux URLs?

Basically—how we auto-detect Linux distributions when we need to and not when we don't want to?

**Comment by jlord** [comment]
Tracing the path from `mongodb-core` → `mongodb-test-runner` → `mongodb-version-manager` → `mongodb-download-url` the only place we're passing in options is `version` in `mongodb-version-manager`: https://github.com/mongodb-js/version-manager/blob/master/index.js#L101. This is of course just one path so I am still not sure if it isn't sure differently (with more options) somewhere else?

**Comment by jlord** [comment]
I've done a strike through of my previous comments because I'm going to push a change to go back to the way it worked before and only detect Linux distributions if an `--evergreen` flag is passed when running `mongodb-test-runner`.

File: index.js [hunk 1]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
...
+function getLinuxDistro() {
+  // TODO we're not checking Debian
+  var formattedDistro = null;
+  var distros = [
+    { '/etc/redhat-release': 'Rhel' },
+    { '/etc/SuSE-release': 'Suse' },
+    { '/etc/lsb-release': 'Ubuntu' }
+  ];
+
+  distros.forEach(function(distro, i) {
```
Comment (daprahamian) [inline index.js]: I'm having trouble following this `forEach`. My intuition is that this should be a `reduce`

File: index.js [hunk 2]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
...
+    }
+  });
+  return formattedDistro;
+}
+
 function parseDistro(opts) {
   if (!opts.distro) {
     if (opts.platform === 'linux') {
       opts.distro = 'linux_' + opts.bits;
+      if (opts.evergreen) opts.linuxDistro = opts.linuxDistro ? opts.linuxDistro : getLinuxDistro(opts);
```
Comment (daprahamian) [inline index.js]: Can we do `= opts.linuxDistro || getLinuxDistro(opts)`?

File: index.js [hunk 3]
```diff
@@ -218,6 +289,16 @@ function resolve(opts, fn) {
           versionId,
           opts.ext
         ].join(''));
+    } else if (opts.platform === 'linux' && opts.evergreen) {
```
Comment (daprahamian) [inline index.js]: Seems like you can do nested `if`s here:

```js
} else if (opts.platform === 'linux') {
  if (opts.evergreen) {
    ...
  } else {
    ...
  }
}
```

File: test/index.test.js [hunk 4]
```diff
@@ -74,6 +74,115 @@ describe('mongodb-download-url', function() {
       verify(done, query,
         'https://fastdl.mongodb.org/linux/mongodb-linux-i686-3.0.7.tgz');
     });
+
+    it('should resolve RHEL 6.2', function(done) {
+      // We support 6.2, and 7.0 community versions
+      var query = {
```
Comment (daprahamian) [inline test/index.test.js]: can these be `const`s?

Comment (jlord) [reply test/index.test.js]: I'm not sure what our rule of thumb should be on updating the style in modules that aren't completely ours as Compass uses this too?

File: index.js [hunk 5]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
...
+    var distroInfo;
+    try {
+      distroInfo = fs.readFileSync(Object.keys(distro)[0]);  // eslint-disable-line
+    } catch (e) {
+      if (i === distros.length - 1) {
+        return formattedDistro;
+      }
+      return;
+    }
+    switch (distros[i][Object.keys(distro)]) {
```
Comment (mbroadst) [inline index.js]: this looks pretty weird, did you expect each of these objects to ever have more than one key?

File: index.js (hunk 1 above)
Comment (mbroadst) [reply index.js]: Agreed with Dan here, I also wonder if you can make this all more streamlined without needing objects or distro titles in the first place. It looks like you could just use a list of "tuples":
```
const distros = [
  ['/etc/redhat-release', parseRhel],
  ['/etc/SuSE-release', parseSuse],
  ['/etc/lsb-release', parseUbuntu]
];

const formattedDistro = distros.reduce((result, distro) => {
  try { 
    return distro[1](fs.readFileSync(distro[0]));
  } catch {
    return null;
  }
});
```

or something along those lines, obviously we could be far more readable with variable names etc :smile:

File: index.js [hunk 6]
```diff
@@ -152,10 +221,12 @@ function parseArch(opts) {
...
   } else if (opts.bits === '32' && opts.platform === 'win32') {
     opts.arch = 'i386';
-  } else {
-    opts.arch = 'x86_64';
-  }
-
+  } else if (opts.arch === 's390x' || opts.arch === 'ppc64le') {
+    opts.enterprise = true;
+  } else if (opts.arch.match('arm')) {
+    // ARM can be Enterprise or community, how do we decide?
```
Comment (mbroadst) [inline index.js]: imho it should be determined based on whether `enterprise === true`, and `enterprise` should be marked `true` IFF `evergreen` is chosen

File: index.js [hunk 7]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
...
+    }
+  });
+  return formattedDistro;
+}
+
 function parseDistro(opts) {
   if (!opts.distro) {
     if (opts.platform === 'linux') {
       opts.distro = 'linux_' + opts.bits;
+      if (opts.evergreen) opts.linuxDistro = opts.linuxDistro || getLinuxDistro(opts);
```
Comment (mbroadst) [inline index.js]: is there a reason we have `linuxDistro` vs `distro`?

File: test/index.test.js (hunk 4 above)
Comment (mbroadst) [reply test/index.test.js]: sure let's leave this as ES5 for now, no need to change everything at once

File: index.js [hunk 8]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
   return opts;
 }
 
+// mongodb-linux-x86_64-rhel62-3.6.2.tgz
+//
+// Red Hat Enterprise Linux Server release 7.2 (Maipo)
+function parseRHEL(data) {
+  if (data.toString().match('7.1')) return 'rhel71';
```
Comment (mbroadst) [inline index.js]: please call `toString` once at the top, otherwise you're converting the buffer to a string with every check here

File: index.js [hunk 9]
```diff
@@ -124,10 +125,78 @@ function parseFileExtension(opts) {
...
+  if (data.toString().match('7.2')) return 'rhel72';
+  if (data.toString().match('6.0')) return 'rhel60';
+}
```
Comment (mbroadst) [inline index.js]: there is no fall through case here, what if none of your matches work? It just returned `undefined`, I think you should be more explicit.

File: index.js (hunk 8 above)
Comment (jlord) [reply index.js]: Realized I'm actually calling `toString` already when passing the argument in so I don't need it here at all 👍

File: index.js (hunk 7 above)
Comment (jlord) [reply index.js]: I saw it as more distinct than the `distro` and but we could append the Linux distribution to `distro` rather than keep it separate.

File: index.js (hunk 6 above)
Comment (jlord) [reply index.js]: Does that mean we'll only test Enterprise builds on Evergreen?

File: index.js (hunk 9 above)
Comment (jlord) [reply index.js]: Good point. I am going to throw an error because it doesn't seem like we should guess and send a version that will possibly not work.

File: index.js [hunk 10]
```diff
@@ -124,10 +125,66 @@ function parseFileExtension(opts) {
...
+    try {
+      return distro[1](fs.readFileSync(distro[0]).toString()); // eslint-disable-line
+    } catch (error) {
+      // Catching file reading error
+      return new Error(error);
+    }
+  }, null);
+  if (formattedDistro instanceof Error) {
+    // Catching parsing version error
+    throw formattedDistro;
```
Comment (jlord) [inline index.js]: @mbroadst @daprahamian This is now switched over to `reduce` but to be honest I think it's messy to twist reduce to work for us in this case. I want to double check that we're ok throwing here?

File: index.js (hunk 6 above)
Comment (mbroadst) [reply index.js]: yes I think that's a fine assumption
